### DISCUSSION TITLE
Bootloader update / ext_csd switching

### DISCUSSION THREAD

**Discussion by nagi-baer** [rauc/rauc#1112] [Q&A]
Hi, I am trying to implement rauc 1.8 into a project (full A/B setup with u-boot in boot0 resp. boot1) and I have problems understanding the rauc status output.
I boot the system and get 

> Booted from: rootfs.0 (A)

as output which refers from my understanding to boot0.
Installing a bundle containing a bootloader and calling the status command again leads to

> Activated: rootfs.1 (B)

which is in my opinion incorrect since the system did not reboot yet. So far so good, but the actual problem is triggering the installation now again. After the second installation trigger the output of the status command is still
 
> Activated: rootfs.1 (B)

but after shutting the device down and restarting, the bootloader from boot0 is run. I would have expected boot1 to be active. Is this a configuration error or is rauc handling an installation twice wrong?

Slot configuration of my system.conf
```
[slot.rootfs.0]
device=/dev/mmcblk2p2
type=raw
bootname=A

[slot.appfs.0]
parent=rootfs.0
device=/dev/mmcblk2p1
type=ext4

[slot.rootfs.1]
device=/dev/mmcblk2p4
type=raw
bootname=B

[slot.appfs.1]
parent=rootfs.1
device=/dev/mmcblk2p3
type=ext4

[slot.bootloader.0]
device=/dev/mmcblk2
type=boot-emmc
```
Thank you in advance!

**Comment by ejoerns** [comment]
> Installing a bundle containing a bootloader and calling the status command again leads to
> 
> > Activated: rootfs.1 (B)
> 
> which is in my opinion incorrect since the system did not reboot yet.

'Activated' means that this is what the bootloader will boot *next*. This of course has to be set *before* actually performing the reboot. By default, this is set after the installation is done.

> After the second installation trigger the output of the status command is still

If you trigger the installation from the same system (without reboot) this is fully valid. The inactive system remains the same as long as the active one does not change.

> but after shutting the device down and restarting, the bootloader from boot0 is run. I would have expected boot1 to be active. Is this a configuration error or is rauc handling an installation twice wrong?

The information you see there has nothing to do with the bootloader you updated. The "Activated" information refers to the slot the bootloader will select.

The atomic bootloader update in RAUC is decoupled from the A/B setup and is visible only in the service logs currently. For the high-level output of RAUC, it just looks like writing and image to any kind of storage. 
I guess this is what caused your confusion here, right? Maybe one could make the extCSD switching more visible, but I am not sure about how, yet.

**Reply by nagi-baer** [reply]
Thank you for your reply. Yes, that confused me. I just checked the ext_csd register using mmc-utils and I realized that the active boot partition is switched after every installation.
What is the reason for decoupling the bootloader from the A/B setup? Our goal is to boot A only with the bootloader in boot0 and vice versa. My current idea is cross checking the ext_csd register and the variable BOOT_ORDER inside u-boot. If they don't match I reconfigure the ext_csd register according to the boot order and reset the board. Is there anything against that approach?

**Reply by ejoerns** [reply]
Normally, strictly coupling the bootloader to the booted system is not desired since this allows atomic updates, but no fallback, etc.
Most update scenarios actually try to prevent bootloader updates whenever possible.

If fallback is not an option and the bootloader must match the booted system for whatever reason, the approach you mentioned can be valid, however.
What's your motivation for couping bootloader update with A/B setup?

**Reply by nagi-baer** [reply]
The board manufacturer releases the kernel and the bootloader always as a bundle. A new kernel release is not necessarily compatible with a previous installed bootloader. In order to prevent booting an invalid constellation the mentioned approach is used. I would implement the fallback logic by putting the logic of the boot script into u-boot.
